### ISSUE TITLE
BeanPropertyWriter / Reader delegation

### ISSUE THREAD

**Issue by shawkins** [FasterXML/jackson-databind#3862]
**Is your feature request related to a problem? Please describe.**

The BeanPropertyWriter and BeanPropertyReader are used to specialize serialization behavior and they have two distinct patterns.

1. Create a completely new instance based upon copying what is passed in like https://github.com/FasterXML/jackson-databind/blob/de62c677b12c02967c9620a11c1c5208f701cfef/src/main/java/com/fasterxml/jackson/databind/ser/impl/UnwrappingBeanPropertyWriter.java#L43 or Jdk8OptionalBeanPropertyWriter.

2. A delegation pattern, such as with https://github.com/FasterXML/jackson-databind/blob/de62c677b12c02967c9620a11c1c5208f701cfef/src/main/java/com/fasterxml/jackson/databind/ser/impl/FilteredBeanPropertyWriter.java#L104

The problem with 1 is that you can only do that style of modification once.  If you introduce multiple serializer modifiers of this style only the last one will win as the subclass of the intermediate one will be lost.

The problem with 2 is that the api is not clearly established for delegation - what set of methods that could possibly affect or use the delegate need to be overwritten?  

We encounter this problem with https://github.com/fabric8io/kubernetes-client/issues/5009

**Describe the solution you'd like**

Based upon the possibility of multiple serialization or deserialization modifiers being applied it does not seem like 1 should be used.  

To enhance the usage of 2 databind should provide a better class / delegation structure here that helps ensure a more predictable experience.  That solution should also be used internally to replace how the UnwrappingBeanPropertyWriter, Jdk8OptionalBeanPropertyWriter, AttributePropertyWriter, etc. are implemented.

**Comment by cowtowncoder** [comment]
I agree with your assessment that approach (1) is problematic -- and I'd go further to say that adding "unwrapping" style was probably a mistake in the first place due to difficulty in implementing and maintaining it. At the same time, however, I am not optimistic this can be changed to unify approaches.

I do agree that all new code should use delegation-based approach.

Having said that, however, any code outside Jackson core (and core modules) should probably try to avoid solutions that rely on either mechanism: they do not compose well.

Will leave this open in case there are some practical improvements, but do not think it likely that unification is possible.
